src/DGtal/base/CSTLAssociativeContainer.h

```diff
@@ -0,0 +1,153 @@
+/**
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+#pragma once
+
+/**
+ * @file CSTLAssociativeContainer.h
+ * @author David Coeurjolly
+ * Laboratoire d'InfoRmatique en Image et Systèmes d'information - LIRIS (CNRS, UMR 5205), CNRS, France
+ *
+ * @date 2015/07/06
+ *
+ * Header file for concept CSTLAssociativeContainer
+ *
+ * This file is part of the DGtal library.
+ */
+
+#if defined(CSTLAssociativeContainer_RECURSES)
+#error Recursive header files inclusion detected in CSTLAssociativeContainer.h
+#else // defined(CSTLAssociativeContainer_RECURSES)
+/** Prevents recursive inclusion of headers. */
+#define CSTLAssociativeContainer_RECURSES
+
+#if !defined CSTLAssociativeContainer_h
+/** Prevents repeated inclusion of headers. */
+#define CSTLAssociativeContainer_h
+
+//////////////////////////////////////////////////////////////////////////////
+// Inclusions
+#include <iostream>
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+  namespace concepts
+  {
+    /////////////////////////////////////////////////////////////////////////////
+    // class CSTLAssociativeContainer
+    /**
+       DescriptionDescription of \b concept '\b CSTLAssociativeContainer' 
+       @ingroup Concepts
+
+       \brief Aim: Defines the concept describing an Associative
+       Container of the STL
+       (https://www.sgi.com/tech/stl/AssociativeContainer.html).
+
+       @note This concept can also be found in the boost library under
+       the name  boost::AssociativeContainer. However, there is a
+       issue in this concept (problem in boost 1.58 but present in
+       several releases) which requires the container to be "sorted". As
+       a consequence, boost::unordered_set does not satisfy the concept.
+
```

Comment:
I agree and this is quite worrying for the future of boost concept check. I am wondering if it will be maintained/evolved in the future by the boost community.